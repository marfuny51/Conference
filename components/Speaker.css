.Table {
    text-align: center;
}

.TableSpeaker {
    border-collapse: collapse;
    width: 70em;
    height: 8em;
    margin: 2em auto 1em auto;
}

.TableSpeakerAdd {
    border-collapse: collapse;
    width: 70em;
    height: 8em;
    margin: 2em auto 1em auto;
    opacity:0.3;
}

.TableHead {
    color: #00D970;
}

.TableHead th {
    font-family: 'Muli', sans-serif;
    background-color: #127a16;
    color: white;
    letter-spacing:0.1em;
    padding: 0.6em;
    border: 0em;
}

.TableBody tr {
    background-color: #d3f1d4;
    border-bottom: 0.1em solid #127a16;
    padding: 0.1em;
    height: 3em;
}

.TableBody tr:hover {
    background-color: #f5f5f5;
}

.TableBody tr td {
    font-family: 'Muli', sans-serif;
    border: 0em;
}

.Read {
    font-family: 'Muli', sans-serif;
    height: 2.2em;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5em;
    border:0.2em solid #77b45b;
}

.Read:hover {
    font-weight: bold;
    height: 2.5em;
    border:0.2em solid #456835;
    cursor: pointer;
}

.Read2 {
    font-family: 'Muli', sans-serif;
    height: 2.2em;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5em;
    border:0.2em solid #77b45b;
    margin-right: 1.2em;
}

.Read2:hover {
    font-weight: bold;
    height: 2.5em;
    border:0.2em solid #456835;
    cursor: pointer;
}

.Add {
    font-family: 'Muli', sans-serif;
    font-size: 1em;
    font-weight: bold;
    height: 2.5em;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5em;
    border:0.2em solid #77b45b;
}

.Add:hover {
    font-size: 1.1em;
    border:0.2em solid #456835;
    cursor: pointer;
}

.NavLink {
    color: #172c11;
    font-weight: bold;
}

.LoadData {
    font-size: 1.1em;
    text-align: center;
    color: white;
}

.AddSpeaker {
    width: 45em;
    height: 13em;
    position: absolute;
    top: 35%;
    left: 32%;
    background-color: #d3f1d4;
    border: 0.2em solid #77b45b;
    border-radius: 0.5em;
}

.Title {
    font-family: 'Muli', sans-serif;
    font-weight: bold;
    letter-spacing:0.1em;
    padding-top: 1em;
}

.Input {
    font-family: 'Muli', sans-serif;
    text-align: left;
    padding: 1em 1em 1em 4em;
}

.Valide {
    color: red;
}

.Textarea {
    width: 20em;
    height: 7em;
}