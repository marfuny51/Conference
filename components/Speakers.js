import React from 'react';
import PropTypes from 'prop-types';
import {voteEvents} from './events';
import isoFetch from 'isomorphic-fetch';
import Speaker from './Speaker';
import AddSpeaker from './addSpeaker';
import EditSpeaker from './editSpeaker';

import './Speaker.css';

class Speakers extends React.PureComponent {
  
  /*constructor(props) {
    super(props);
    this.mounted = false;
  }*/

  state = {
    dataReady: false,
    speakers: [],
    className: 'TableSpeaker',
    editCode: null,
    mode: null, //0 -view, 1- edit, 2 -add
  }

  speakersArray;

  componentDidMount = () => {
    this.loadData();
    //this.mounted = true;
    voteEvents.addListener('ESave',this.speakerSave);
    voteEvents.addListener('ECancel',this.speakerCancel);
    voteEvents.addListener('EDelete',this.delete);
    voteEvents.addListener('EEdit',this.editSpeaker);
    voteEvents.addListener('EditSave',this.speakerSave);     
  };

  componentWillUnmount = () => {
    //this.mounted = false;
    voteEvents.removeListener('ESave',this.speakerSave);
    voteEvents.removeListener('ECancel',this.speakerCancel);
    voteEvents.removeListener('EDelete',this.delete);
    voteEvents.removeListener('EEdit',this.editSpeaker);
    voteEvents.removeListener('EditSave',this.speakerSave);
  };

  loadData = () => {
    let ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    let sp = new URLSearchParams();
    sp.append('f', 'READ');
    sp.append('n', 'PROBA_PERA7');

    isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp,
    })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result === "") {
              this.speakersArray = {};
              this.speakersArray.speakers = [];
              this.speakersArray.topics = [];
              this.speakersArray.members = [];
            }
            else this.speakersArray = JSON.parse(data.result);
            console.log(this.speakersArray.speakers);
            
              this.setState({ dataReady:true, speakers:this.speakersArray.speakers, className:'TableTopic'});
            
        })
        .catch( error => {
            this.fetchError(error.message);
        });

  }; 

  fetchError = (errorMessage) => {
    console.error(showStr);
  };

  speakerSave = (id, name, phone, position, topic) => {
    var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    var updatePassword=Math.random();
    let sp1 = new URLSearchParams();
    sp1.append('f', 'LOCKGET');
    sp1.append('n', 'PROBA_PERA7');
    sp1.append('p', updatePassword);
    
    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp1,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( data => {
        if (data.result.speakers === undefined||data.result.speakers==='') {
          this.speakersArray.speakers =[];
        }
        else this.speakersArray.speakers = JSON.parse(data.result.speakers);
      })
      .catch( error => {
          console.log(error.message);
      });
    
      if (this.state.mode === 1) {
        console.log(name, phone, position, topic);
        this.speakersArray.speakers.forEach( speaker => {
          if (speaker.id === id) {
            speaker.name = name;
            speaker.phone = phone;
            speaker.position = position;
            speaker.topic = topic;
          }
        })
      }

      else if (this.state.mode === 2) {
        this.speakersArray.speakers.push({id: id, name: name, phone: phone, position: position, topic: topic});
        console.log(this.speakersArray);
      }

    let sp2 = new URLSearchParams();
    sp2.append('f', 'UPDATE');
    sp2.append('n', 'PROBA_PERA7');
    sp2.append('p', updatePassword);
    sp2.append('v', JSON.stringify(this.speakersArray));

    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp2,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( (data) => {
        console.log(data);
        this.loadData();
        this.setState({mode:0});        
      })
      .catch( error => {
          this.fetchError(error.message);
      });
    }

    delete = (id) => {
      var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
      var updatePassword=Math.random();
      let sp3 = new URLSearchParams();
      sp3.append('f', 'LOCKGET');
      sp3.append('n', 'PROBA_PERA7');
      sp3.append('p', updatePassword);
      
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp3,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result.speakers === undefined||data.result.speakers==='') {
              this.speakersArray.speakers =[];
            }
            else this.speakersArray.speakers = JSON.parse(data.result.speakers);
        })
        .catch( error => {
            console.log(error.message);
        });
      
      this.speakersArray.speakers = this.speakersArray.speakers.filter(speaker => speaker.id !== id);
      console.log(this.speakersArray);
  
      let sp2 = new URLSearchParams();
      sp2.append('f', 'UPDATE');
      sp2.append('n', 'PROBA_PERA7');
      sp2.append('p', updatePassword);
      sp2.append('v', JSON.stringify(this.speakersArray));
  
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp2,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( (data) => {
            console.log(data);
            this.loadData();
        })
        .catch( error => {
            this.fetchError(error.message);
        });
      }
  

  speakerCancel = () => {
    let speakers = [...this.state.speakers];
    this.setState({mode:0, speakers: speakers, className:'TableTopic'});
  }

  addSpeaker = (EO) => {
    this.setState({mode:2, className:'TableTopicAdd'})
  }
  
  editSpeaker = (id) => {
    this.setState( {editCode:id, mode:1, className:'TableTopicAdd'})
  }

  getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
  
  render() {
    if ( !this.state.dataReady )
      return <div className = 'LoadData'>загрузка данных...</div>;

    let speakers = [...this.state.speakers];
    let speakers2 = [...this.state.speakers];

    var speakersCode=speakers.map( speaker =>
      <Speaker key={speaker.id} 
      mode={this.state.mode}
      speaker={speaker}  />
    );
    var editCode = speakers2.find( speaker => speaker.id === this.state.editCode); 

    return (
      <div className='Table'>
        <table className={this.state.className}>
        <colgroup>
        <col className="cols1"/>
        <col span="2" className="cols2"/>
        <col className="cols3"/>
        <col span="2" className="cols4"/>
        </colgroup>
        <thead className='TableHead'>
              <tr>
                  <th>Name</th>
                  <th>Phone</th>
                  <th>Position</th>
                  <th>Topic</th>
                  <th>Edit</th>
                  <th>Delete</th>
              </tr>
          </thead>
          <tbody className='TableBody'>
          {speakersCode}
          </tbody>
        </table>
        <input type="button" value="Add a new speaker" className='Add' onClick = {this.addSpeaker} disabled = {(this.state.mode===2)?true:false}/>

      {
        (this.state.mode===2)&&
        <AddSpeaker key={this.getRandomInt(1, 10000)}
        mode={this.state.mode}
        id={this.getRandomInt(1, 10000)}
        name=''
        phone=''
        position=''
        topic=''/>
      }

      {
        (this.state.mode===1)&&
        <EditSpeaker key={editCode.id}
        mode={this.state.mode}
        id={editCode.id}
        name= {editCode.name}
        phone={editCode.phone}
        position={editCode.position}
        topic={editCode.topic}/>
      }

      </div>
    )
    ;

  }

}

export default Speakers;
