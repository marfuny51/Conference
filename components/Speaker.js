import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import './Speakers.css';

class Speaker extends React.PureComponent {

  static propTypes = {
    speakers:PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
          phone: PropTypes.string.isRequired,
          position: PropTypes.string.isRequired,
          topic: PropTypes.string.isRequired,
        })
      ),
  };
  
  render() {

    return (
    <tr key={this.props.info.id}>
        <td><NavLink to={"/speaker/"+this.props.info.name}>{this.props.info.name}</NavLink></td>
        <td>{this.props.info.phone}</td>
        <td>{this.props.info.position}</td>
        <td>{this.props.info.topic}</td>
    </tr>
    );

  }

}

export default Speaker;
