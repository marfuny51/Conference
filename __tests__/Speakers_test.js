import React from 'react'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json'
import Speaker from '../components/Speaker'
import Speakers from '../components/Speakers'

Enzyme.configure({ adapter: new Adapter() });

const component = shallow(<Speakers />);
const speakers = [{id:460,name:"Inokentii Zhivago",phone:"+375442569847",position:"Business analyst",topic:"Business analysis in software development"},{id:1348,name:"Veronika Dokuro",phone:"+375332548965",position:"Q&A Engineer",topic:"Software testing"}];
describe('<Speakers /> component', () => {
    it('should render', () => {
        component.setState({ dataReady: true, speakers: speakers});
        speakers.map( speaker =>
            shallow(<Speaker key={speaker.id} 
                speaker={speaker}/>)
          );
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();

        component.find('.Add').simulate("click");
    });
});