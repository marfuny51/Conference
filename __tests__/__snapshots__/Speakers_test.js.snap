// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Speakers /> component should render 1`] = `
<div
  className="Table"
>
  <table
    className="TableSpeaker"
  >
    <colgroup>
      <col
        className="cols1"
      />
      <col
        className="cols2"
        span="2"
      />
      <col
        className="cols3"
      />
      <col
        className="cols4"
        span="2"
      />
    </colgroup>
    <thead
      className="TableHead"
    >
      <tr>
        <th>
          Name
        </th>
        <th>
          Phone
        </th>
        <th>
          Position
        </th>
        <th>
          Topic
        </th>
        <th>
          Edit
        </th>
        <th>
          Delete
        </th>
      </tr>
    </thead>
    <tbody
      className="TableBody"
    >
      <Speaker
        key="460"
        mode={null}
        speaker={
          Object {
            "id": 460,
            "name": "Inokentii Zhivago",
            "phone": "+375442569847",
            "position": "Business analyst",
            "topic": "Business analysis in software development",
          }
        }
      />
      <Speaker
        key="1348"
        mode={null}
        speaker={
          Object {
            "id": 1348,
            "name": "Veronika Dokuro",
            "phone": "+375332548965",
            "position": "Q&A Engineer",
            "topic": "Software testing",
          }
        }
      />
    </tbody>
  </table>
  <input
    className="Add"
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Add a new speaker"
  />
</div>
`;
